collaborative-plonk/src/round3.rs

```diff
@@ -0,0 +1,602 @@
+
+            let mut pi = T::zero_share();
+            for (j, lagrange) in zkey.lagrange.iter().enumerate() {
+                let l_eval = lagrange.evaluations[i];
+                let a_val = T::index_sharevec(&polys.buffer_a, j);
+                let tmp = driver.mul_with_public(&l_eval, &a_val);
+                pi = driver.sub(&pi, &tmp);
+            }
+
+            e1_ = driver.add(&e1_, &pi);
+            e1_ = driver.add_with_public(&qc, &e1_);
+            e1.push(e1_);
+            e1z.push(e1z_);
+
+            let betaw = challenges.beta * w;
+            let mut e2a_ = a.clone();
+            e2a_ = driver.add_with_public(&betaw, &e2a_);
+            e2a_ = driver.add_with_public(&challenges.gamma, &e2a_);
+            e2a.push(e2a_);
+
+            let mut e2b_ = b.clone();
+            e2b_ = driver.add_with_public(&(betaw * zkey.verifying_key.k1), &e2b_);
+            e2b_ = driver.add_with_public(&challenges.gamma, &e2b_);
+            e2b.push(e2b_);
+
+            let mut e2c_ = c.clone();
+            e2c_ = driver.add_with_public(&(betaw * zkey.verifying_key.k2), &e2c_);
+            e2c_ = driver.add_with_public(&challenges.gamma, &e2c_);
+            e2c.push(e2c_);
+
+            let e2d_ = z;
+            e2d.push(e2d_);
+
+            let mut e3a_ = a;
+            e3a_ = driver.add_with_public(&(s1 * challenges.beta), &e3a_);
+            e3a_ = driver.add_with_public(&challenges.gamma, &e3a_);
+            e3a.push(e3a_);
+
+            let mut e3b_ = b;
+            e3b_ = driver.add_with_public(&(s2 * challenges.beta), &e3b_);
+            e3b_ = driver.add_with_public(&challenges.gamma, &e3b_);
+            e3b.push(e3b_);
+
+            let mut e3c_ = c;
+            e3c_ = driver.add_with_public(&(s3 * challenges.beta), &e3c_);
+            e3c_ = driver.add_with_public(&challenges.gamma, &e3c_);
+            e3c.push(e3c_);
+
+            let e3d_ = zw;
+            e3d.push(e3d_);
+            w *= pow_plus2_root_of_unity;
+        }
+
+        let e2a_vec = e2a.into();
+        let e2b_vec = e2b.into();
+        let e2c_vec = e2c.into();
+        let e2d_vec = e2d.into();
+        let zp_vec = zp.into();
+
+        let [e2, e2z_0, e2z_1, e2z_2, e2z_3] = mul4vec!(
+            driver, &e2a_vec, &e2b_vec, &e2c_vec, &e2d_vec, &ap_vec, &bp_vec, &cp_vec, &zp_vec,
+            &domain1
+        );
+
+        let e3a_vec = e3a.into();
+        let e3b_vec = e3b.into();
+        let e3c_vec = e3c.into();
+        let e3d_vec = e3d.into();
+        let zwp_vec = zwp.into();
+
+        let [e3, e3z_0, e3z_1, e3z_2, e3z_3] = mul4vec!(
+            driver, &e3a_vec, &e3b_vec, &e3c_vec, &e3d_vec, &ap_vec, &bp_vec, &cp_vec, &zwp_vec,
+            &domain1
+        );
+
+        let mut t_vec = Vec::with_capacity(zkey.domain_size * 4);
+        let mut tz_vec = Vec::with_capacity(zkey.domain_size * 4);
+        for i in 0..zkey.domain_size * 4 {
```

Comment:
same for this loop

Reply:
commented and enforced with for_each loop